packages/salesforce-adapter/src/change_validators/unknown_users.ts

```
@@ -0,0 +1,138 @@
+  ChangeError, isAdditionOrModificationChange, isInstanceChange, ChangeValidator,
+  InstanceElement, getChangeData,
+} from '@salto-io/adapter-api'
+import { apiName } from '../transformers/transformer'
+import { isInstanceOfType, buildSelectQueries, queryClient } from '../filters/utils'
+import SalesforceClient from '../client/client'
+
+const { awu } = collections.asynciterable
+
+// cf. 'Statement Character Limit' in https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_select.htm
+const SALESFORCE_MAX_QUERY_LEN = 100000
+
+type UserGetterFunc = (instance: InstanceElement) => string | undefined
+
+type FieldGetterDef = {
+  field: string
+  getter: UserGetterFunc
+}
+
+type UserGetterDef = {
+  [type: string]: FieldGetterDef[]
+}
+
+type MissingUserInfo = {
+  instance: InstanceElement
+  field: string
+  userName: string
+}
+
+const getCaseSettingsOwner = (instance: InstanceElement,): string | undefined => {
+  if (instance.value.defaultCaseOwnerType !== 'User') {
+    return undefined
+  }
+  return instance.value.defaultCaseOwner
+}
+
+const USER_GETTERS: UserGetterDef = {
+  CaseSettings: [
+    {
+      field: 'defaultCaseUser',
+      getter: instance => instance.value.defaultCaseUser,
+    },
+    {
+      field: 'defaultCaseOwner',
+      getter: instance => getCaseSettingsOwner(instance),
+    },
+  ],
+}
+
+const isValidUser = (user: string|undefined): user is string => !_.isUndefined(user)
+
+const getUsersFromInstance = (instance: InstanceElement, getterDefs: FieldGetterDef[]): string[] => (
+  getterDefs.map(getterDef => getterDef.getter(instance)).filter(isValidUser)
+)
+
+const getUsersFromInstances = async (
+  defMapping: UserGetterDef,
+  instances: InstanceElement[]): Promise<string[]> => (
+  awu(instances).map(async instance => {
+    const instanceType = await apiName(await instance.getType())
+    const getterDefs = defMapping[instanceType]
+    return getUsersFromInstance(instance, getterDefs)
```

Is this code type safe? 
(aka, if line 78 now returns `acmeCorp` will line 79 & 80 still properly work?)

---

Before we call this function we filter the input on whether its type appears in `USER_GETTERS` so I don't have to code defensively in every function I call. Do you feel this makes the function too special-purpose?

---

I hear you. That's why I said that this code is not type-safe(and I did not say that the code itself is not safe).

Ideally, when I write code, I try to make sure my assumptions are refelcted in the types, to utilize the fact that the compiler can verify them for me in compile-time.
And in this case that's not true. (aka, the type-system allows line 77 to have value `acmeCorp` and line 78 assumes that it always exists in defMapping).

Can we make sure your assumption & filter, are refelected in the types that are passed to this function?

(Worth to say: this is probably a nit)

---

I get what you're saying and I agree it would be better, but I'm not familiar enough with the Typescript type system and I'm not sure how I would do that considering apiName() returns a generic string. Could use some guidance on this.

---

Here is a snippet of what I would do here 

```
/*
*                      Copyright 2022 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { collections } from '@salto-io/lowerdash'
import {
  ChangeError, isAdditionOrModificationChange, isInstanceChange, ChangeValidator,
  InstanceElement, getChangeData,
} from '@salto-io/adapter-api'
import { apiName } from '../transformers/transformer'
import { isInstanceOfType, buildSelectQueries, queryClient } from '../filters/utils'
import SalesforceClient from '../client/client'

const { awu } = collections.asynciterable

// cf. 'Statement Character Limit' in https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_select.htm
const SALESFORCE_MAX_QUERY_LEN = 100000
type FieldName = 'defaultCaseUser' | 'defaultCaseOwner'
type GetUserField = (instance: InstanceElement, field: FieldName) => string | undefined
type UserFieldGetter = {
  field: FieldName
  getter: GetUserField
}
const TypeWithField = {
  CaseSettings: 'CaseSettings',
} as const
type TypeWithField = keyof typeof TypeWithField
const toTypeWithField = (str: string): TypeWithField | undefined => Object.values(TypeWithField).find(x => x === str)
type TypesWithUserFields = Record<TypeWithField, UserFieldGetter[]>

type MissingUser = {
  instance: InstanceElement
  field: string
  userName: string
}

const getCaseSettingsOwner = (instance: InstanceElement, fieldName: string): string | undefined => {
  if (instance.value?.[fieldName] !== 'User') {
    return undefined
  }
  return instance.value?.[fieldName]
}

const USER_GETTERS: TypesWithUserFields = {
  CaseSettings: [
    {
      field: 'defaultCaseUser',
      getter: (instance, field) => instance.value?.[field],
    },
    {
      field: 'defaultCaseOwner',
      getter: (instance, field) => getCaseSettingsOwner(instance, field),
    },
  ],
}

const isValidUser = (user: string|undefined): user is string => !!user

const getUsersFromInstance = (instance: InstanceElement, getterDefs: UserFieldGetter[]): string[] => (
  getterDefs.map(({ getter, field }) => getter(instance, field)).filter(isValidUser)
)

const getUsersFromInstances = async (
  defMapping: TypesWithUserFields,
  instances: InstanceElement[]): Promise<string[]> => (
  awu(instances).map(async instance => {
    const instanceType = await apiName(await instance.getType())
    const typeWithFields = toTypeWithField(instanceType)
    const getterDefs = typeWithFields ? defMapping[typeWithFields] : []
    return getUsersFromInstance(instance, getterDefs)
  }).flat().toArray()
)

const getSalesforceUsers = async (client: SalesforceClient, users: string[]): Promise<string[]> => {
  if (users.length === 0) {
    return []
  }

  const queries = await buildSelectQueries('User', ['Username'], users.map(userName => ({ Username: userName })), SALESFORCE_MAX_QUERY_LEN)

  return awu(await queryClient(client, queries)).map(sfRecord => sfRecord.Username).toArray()
}

const getUnknownUsers = async (
  defMapping: TypesWithUserFields,
  instance: InstanceElement,
  existingUsers: string[],
): Promise<MissingUser[]> => {
  const instanceType = await apiName(await instance.getType())
  const typeWithFields = toTypeWithField(instanceType)

  const getterDefs = typeWithFields ? defMapping[typeWithFields] : []

  return (getterDefs
    .map(({ getter, field }) => ({ instance, field, userName: getter(instance, field) }))
    .filter(missingUserInfo => isValidUser(missingUserInfo.userName)) as MissingUser[])
    .filter(missingUserInfo => !existingUsers.includes(missingUserInfo.userName))
}

const unknownUserError = ({ instance, field, userName }: MissingUser): ChangeError => (
  {
    elemID: instance.elemID,
    severity: 'Error',
    message: `The field ${field} in '${instance.elemID.getFullName()}' refers to the user '${userName}' which does not exist in this Salesforce environment`,
    detailedMessage: `The field ${field} in '${instance.elemID.getFullName()}' refers to the user '${userName}' which does not exist in this Salesforce environment`,
  }
)

/**
 * Fields that reference users may refer to users that don't exist. The most common case would be when deploying
 * between different environment, as users by definition can't exist in multiple environments.
 */
const changeValidator = (client: SalesforceClient): ChangeValidator => async changes => {
  const instancesOfInterest = await awu(changes)
    .filter(isAdditionOrModificationChange)
    .filter(isInstanceChange)
    .map(getChangeData)
    .filter(isInstanceOfType(...Object.keys(USER_GETTERS)))
    .toArray()

  const userRefs = await getUsersFromInstances(USER_GETTERS, instancesOfInterest)
  const existingUsers = await getSalesforceUsers(client, userRefs)

  return awu(instancesOfInterest)
    .flatMap(async elem => getUnknownUsers(USER_GETTERS, elem, existingUsers))
    .map(unknownUserError)
    .toArray()
}

export default changeValidator

```

---

Also note:

1.  You can remove the filter `filter(isInstanceOfType(...Object.keys(USER_GETTERS)))` as it's not needed
2. The filter before caused a dependency between `getUsersFromInstances` and it's caller - which is not ideal - aka, if someone were to use this function from somewhere else, he might provide it a different set of instances, that were not filtered, or if someone were to change the filter, in this code, he might not notice(and it's really hard to ) that the `getUsersFromInstances` depends on it ...

---

OK, so the essence here is 
```
const TypeWithField = {
  CaseSettings: 'CaseSettings',
} as const
type TypeWithField = keyof typeof TypeWithField
const toTypeWithField = (str: string): TypeWithField | undefined => Object.values(TypeWithField).find(x => x === str)
```

---

```
const getCaseSettingsOwner = (instance: InstanceElement, fieldName: string): string | undefined => {
  if (instance.value?.[fieldName] !== 'User') {
    return undefined
  }
  return instance.value?.[fieldName]
}

const USER_GETTERS: TypesWithUserFields = {
  CaseSettings: [
    {
      field: 'defaultCaseUser',
      getter: (instance, field) => instance.value?.[field],
    },
    {
      field: 'defaultCaseOwner',
      getter: (instance, field) => getCaseSettingsOwner(instance, field),
    },
  ],
}
```
This will (if I understand it correctly) make the functions fail silently, which I think is actually a pretty bad outcome. I'll modify this a bit to at least log something.